pkg/reconciler/buildrun/buildrun.go

```diff
@@ -416,91 +428,78 @@ func (r *ReconcileBuildRun) Reconcile(ctx context.Context, request reconcile.Req
+			if err := resources.DeleteServiceAccount(ctx, r.client, buildRun); err != nil {
+				ctxlog.Error(ctx, err, "Error during deletion of generated service account.")
 				return reconcile.Result{}, err
 			}
+		}
+		objName := buildRunner.GetObject().GetName()
+		buildRun.Status.TaskRunName = &objName
 
-			resources.UpdateBuildRunUsingTaskFailures(ctx, r.client, buildRun, lastTaskRun)
-			taskRunStatus := trCondition.Status
-
-			// check if we should delete the generated service account by checking the build run spec and that the task run is complete
-			if taskRunStatus == corev1.ConditionTrue || taskRunStatus == corev1.ConditionFalse {
-				if err := resources.DeleteServiceAccount(ctx, r.client, buildRun); err != nil {
-					ctxlog.Error(ctx, err, "Error during deletion of generated service account.")
-					return reconcile.Result{}, err
-				}
-			}
-
-			buildRun.Status.TaskRunName = &lastTaskRun.Name
+		if origStatus.StartTime == nil && buildRun.Status.StartTime != nil {
+			// Report the buildrun established duration (time between the creation of the buildrun and the start of the buildrun)
+			buildmetrics.BuildRunEstablishObserve(
+				buildRun.Status.BuildSpec.StrategyName(),
+				buildRun.Namespace,
+				buildRun.Spec.BuildName(),
+				buildRun.Name,
+				buildRun.Status.StartTime.Time.Sub(buildRun.CreationTimestamp.Time),
+			)
+		}
 
-			if buildRun.Status.StartTime == nil && lastTaskRun.Status.StartTime != nil {
-				buildRun.Status.StartTime = lastTaskRun.Status.StartTime
+		if origStatus.CompletionTime == nil && buildRun.Status.CompletionTime != nil {
 
-				// Report the buildrun established duration (time between the creation of the buildrun and the start of the buildrun)
-				buildmetrics.BuildRunEstablishObserve(
-					buildRun.Status.BuildSpec.StrategyName(),
-					buildRun.Namespace,
-					buildRun.Spec.BuildName(),
-					buildRun.Name,
-					buildRun.Status.StartTime.Time.Sub(buildRun.CreationTimestamp.Time),
-				)
-			}
+			// buildrun completion duration (total time between the creation of the buildrun and the buildrun completion)
+			buildmetrics.BuildRunCompletionObserve(
+				buildRun.Status.BuildSpec.StrategyName(),
+				buildRun.Namespace,
+				buildRun.Spec.BuildName(),
+				buildRun.Name,
+				buildRun.Status.CompletionTime.Time.Sub(buildRun.CreationTimestamp.Time),
+			)
 
-			if lastTaskRun.Status.CompletionTime != nil && buildRun.Status.CompletionTime == nil {
-				buildRun.Status.CompletionTime = lastTaskRun.Status.CompletionTime
+			podCreationTime := buildRunner.GetPodCreationTime(ctx, r.client)
+			podRampUpTime := buildRunner.GetPodInitFinishedTime(ctx, r.client)
 
-				// buildrun completion duration (total time between the creation of the buildrun and the buildrun completion)
-				buildmetrics.BuildRunCompletionObserve(
+			if podCreationTime != nil {
+				// taskrun ramp-up duration (time between taskrun creation and taskrun pod creation)
+				buildmetrics.TaskRunRampUpDurationObserve(
```

Comment:
Similar issue - this metric is tied to an underlying implementation. In theory a runner's API does not need to create Kubernetes pods, it just needs to represent the execution of a workload. For example, a KubeVirt `VirtualMachineInstance`: https://kubevirt.io/user-guide/user_workloads/lifecycle/